osm2streets/src/lib.rs

```diff
@@ -286,6 +303,60 @@ impl StreetNetwork {
                 .push((self.roads[&r].untrimmed_road_geometry().0, label.into()));
         }
     }
+
+    // Restore the invariant that an intersection's roads are ordered clockwise
+    //
+    // TODO This doesn't handle trim_roads_for_merging
```

Comment:
In cases where we remove a short road and merge intersections, sorting around a "center" point gets kind of funny. `geometry/algorithm.rs` takes care to try and handle that. I didn't port the same logic here yet.

Reply:
I think the best way to determine the order of roads in those multi-node intersections from scratch is to use the OSM graph structure. First sorting the nodes (that are the Road endpoints) then by bearing of the first OSM segment.  https://github.com/a-b-street/osm2streets/pull/98#issuecomment-1259035988

This would require keeping the OSM graph around in some capacity, at least within multi-node intersections.

Reply:
I agree with keeping the graph structure around somehow. Preserving turn restrictions when we merge gets very complex, and a possibly simpler approach is to use the original graph and just ask questions about it. Keeping the graph also means we can map back to objects in OSM more directly.

I don't know how this would work, so shelving the idea for the moment, but you've definitely got me thinking...

Reply:
I like "Asking questions about it". In fact, load and leave the osm graph completely in place. Every road should get it's own copy of the geom (like when you merge similar ways currently) which can be promptly offset to account for placement. Then further tweaked if needed (like sausage links). (Think of an editor, where the graph is literally drawn on the map. One could point at a lane to edit the way.)

(Roads and intersections need to have IDs independant of the OSM ids they reference, [and they need to know their own ID, so we don't need to pass pairs eveywhere])

Asking questions about roads is how I'm thinking about movements too. It's like I want to iterate flows of traffic clockwise around the intersection. A way could yield a single (`oneway`) flow, or commonly two. An intersection provides an `iter_cw_flows() -> Flow`, where Flow is a view onto a subset of lanes on a Road (that all go in the same direction).

I think that's how the type system helps me out here: There exists a whole lot of data about a whole lot of lanes, laid out in memory in one way or another, I want to slice it up differently, with differently typed structs, for different tasks. It's like the [HashMap Entry type](https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html).

Keeping an immutable copy of the OSM graph and making the Street graph independent will no doubt become GH Issues to discuss and target soon.